user-initiated sharing

minrk opened the pull request:
Implements the possibility of users granting each other permission to access their servers.

Summary of the design for review:

- sharing is disabled by default
- sharing can be enabled by granting the `shares` permission to all or a subset of users
- there is no sharing UI, only an API
- sharing can be explicit via user or group name (requires `read:users:name` or `read:groups:name` permission in order to grant specific access)
- sharing can be granted via "share code", which only requires the `shares` permission, no other permissions
- share codes must expire, and expire after 24 hours by default
- share codes are exchanged for updating the Share for the given (user, server) combination
- users can review/remove themselves from shares with the `users:shares` scope (part of default user scopes)
- A user's permission on a server is always represented by exactly one Share row in the database. Accepting more shares updates this row, which makes expiration potentially confusing, so Shares do not currently have a way to expire.

Questions to consider:

- default permissions (off by default, seems clear)
- separate permissions for codes/explicit shares? It's currently not possible to enable sharing by name without also enabling sharing via code
- limiting sharing? There is no mechanism by which to limit who a given user can share _with_ (via code, at least). A limitation on `read:users:name` limits who they can share with by name, but there is not a mechanism to limit sharing by name if the user has other reasons to have `read:users:name`
- there is no consent confirmation step for sharing by name. That means if you have scopes `shares,read:users:name!user=xyz`, then you can grant `xyz` access to your server without their participation at all. This is a major avenue for abuse and spam in Google Docs / Calendar, but I'm not sure that applies to JupyterHub? There's nothing you can _do_ with a shared server once you grant the permission until the shared-with user visits the server and proceeds through the oauth confirmation steps.
- Expiring Shares themselves. This is complicated by the fact that there is at most one Share for a given (user, server). What should happen if a user has expiring permissions and is then granted different permissions with different expiry? The current choices are: ignore updated expiration (surprising) or override old expiration (also surprising). Current solution is to avoid the situation by not expiring Shares at all, which matches ~all sharing I've ever seen (e.g. GitHub, Google Docs, slack, etc.) where access must be revoked explicitly once granted.

Scope limitations:

- Shares are per-server, so you cannot use sharing to grant access to all of a user's named servers via sharing
- Shares do not expire
- Shares cannot grant access to anything except for 

----

##  Original PR description below

Absolutely not close to working or ready for review (I'll be rewriting history a bunch), but opening as a placeholder since folks are asking about it.

The closest to ready is the new [doc explaining how sharing is meant to work](https://github.com/minrk/jupyterhub/blob/user-shares/docs/source/tutorial/sharing.md).


The idea is that a 'share' is a granting of permission (usually access) associated with a specific server to 

shares are essentially mini roles, which always grant permissions restricted to a single server (usually just access:servers!server, but potentially including custom scopes, e.g. for read-only access). See #4349 for my thoughts on why I think this is probably going to work better than the more powerful but much more complex full user-initiated roles in #3858.

## A Share

- grants permission(s) for ONE server to ONE user OR group
- record who created them
- can expire

## And users should be able to:

- grant access to their servers (create Shares)
- revoke access to their servers (delete Shares)
- see who has access to their servers (list Shares by server / owner)
- see what's been shared with them (list Shares by assigned user/group)

Managing sharing permissions is tied only to the server being shared, there is no limit placed on _who_ it can be shared with. This also means an admin can grant access and the user can revoke that same access.

## Questions:

- should users be able to delete "shared with me?"
- do we need a single 'list all the shares for everyone' endpoint? I started to implement this, but the permission filter is complicated (we have a similar implementation in listing users and groups, so it's doable, but nice to avoid
- how do we address leaking user and group names via sharing (i.e. if i try to share with `xyz`, I can know whether there is a user with that name. Is that a problem?)?
- should users have permission to share access to their servers by default, or should it be opt-in for admins (i.e. in the default `user` role?

I'd like to start as minimal as possible while still being actually useful.

As with other things, I plan to implement this as 'api-only' first, and UI can be its own task. Sharing UI is hard, and I'd like to try it in an extension first.

closes #4349

minrk commented:
I've got some time to put into this again before getting buried in the end of the semester, and I'm wondering if the `created_by` field is useful. Right now, I track who _created_ a sharing permission, and I was thinking of reviewing permissions that you've granted, plus. But this isn't really something I see elsewhere, and as soon as _modifying_ permissions starts to enter into it, it becomes super complicated.

Removing `created_by` from the db and leaving it to logging for audit purposes simplifies things, and then the "permissions granted by" question becomes one for logs, not for the db itself.

Another question is about _modifying_ permissions. Normally, permissions are a single dict per user (i.e. a _user_ has X permissions) and don't expire. But as written, a Server can be shared with a User more than once with different permissions and different expiration. This enables the possibility of e.g. "alice has read-only permissions to bob's server, but bob can grant alice elevated permissions for one hour, which will then revert to before."

Again, this is a weird situation, and I don't really like being in the business of weird situations, permissions-wise. So I see two main options in terms of permissions _modification_:

1. modify a single record of permissions for a single user on a single server, which makes expiration weird
2. allow multiple permission grants per user, which makes modifications like revoking permissions odd, but expiration of bundles of permissions sensible

So it seems to me that permissions should be a single record per `(user, server)` combination, and 'expiration' is purely a single number for expiration of _all_ access to that server. The one weird case I can see for that is if a user _already_ has permanent access to a server with permission e.g. `access:servers!server` and then accepts temporary elevated permissions `admin:servers!server` with an expiration date, what should happen? I see 3 choices:

1. elevated permissions are granted, but permanently (exceeds request)
2. elevated permissions are granted and _all_ permissions expire (after share, permissions are _less_ than before)
3. error - if you already have permissions granted, you can't accept an invitation to modify your permissions, permissions for `(user, server)` must be modified explicitly via the API

Expiration is really the only thing that complicates this, because otherwise they are just additions/subtractions from a list. I'm thinking 3 makes the most sense, because there's no guessing involved. It's also not common for permissions to expire.

Finally, one last option is a bit in-between, but more complex to implement, which is to remove the concept of expiration from _user_ permissions entirely, and only apply them to **invitation codes**. I.e. I can use the InviteCode approach to share permissions, and _those_ permissions can expire, and they are tied to the expiration of the invite code. But _only_ permissions granted that way. Right now, I have invitations as a temporary mechanism that exchanges a code for a real change in permissions, but once exchanged, there is no relationship between the code and the permissions granted (i.e. it's exactly the same as sharing with a specific user, you just didn't need to know the username beforehand). Changing this, such that permissions-from-invitations are more substantially different from permissions-granted-directly, would I think make expiration sensible (it makes a bit of sense to be 'invited' a few times), such that when a code expires or is revoked, all permissions granted by that code are also revoked.

I really want to make this first version as simple as possible, but combining invite codes and permission expiration is making this hairier than I'd like. I think it should be on the table to drop expiration of permissions entirely. It's quite a rare feature, and much more common for _tokens_ than users.

Yet another possible simplification is to make the invite _code_ singular per server, again matching Google Docs - where "people with the link" permissions are singular, and can't vary (they also don't expire, but can be revoked).

So here's a proposal that I think simplifies things, assuming we want to preserve the desire for sharing to expire:

1. a 'share' can be with a user, group, or 'code'
2. there can be only one 'share' for a given server per user, group, or code
3. _only_ codes can expire, so if you want to share permissions temporarily, you must use a code
4. when a code expires, all permissions granted _via_ that code are revoked

I'm on the fence about whether a given server should be allowed to have more than one code at a time. It makes sense structurally and in principle, but again this is something I haven't seen elsewhere and I don't like being in the business of unique things permissions-wise.

manics commented:
I agree with starting with something simple, especially if it's designed so that other features can be added without breaking changes.

Token sharing with automatic expiry is nice, how much demand is there for it? Is it essential for a first release?

Looking at the overall plan for sharing, does all of it need to be in core JupyterHub or could some of it be done in extensions/services?

minrk commented:
> Looking at the overall plan for sharing, does all of it need to be in core JupyterHub or could some of it be done in extensions/services?

I think it needs to be in Hub, because there is currently no mechanism by which an extension can grant permissions at runtime.

If I were to make the _simplest_ user-initiated sharing proposal, it wouldn't include expiration:

- share codes are exchanged for permissions, but not linked after permissions are granted (i.e. a share-code is identical to granting permission to a user by name, it's just letting the shared-with user enter their own name)
- single permissions per `(user, server)` pair
- share codes expire (usually short-lived), but permissions do not

That is: permissions are permanent (can be revoked, but do not expire automatically), and sharing via code can _permanently expand_ a user's permissions, but not grant anything temporarily. I believe that solves most of the challenges. The other simpler alternative is:

- single permissions per `(user, server)` pair (same)
- _single_ share code with single permissions (Google Docs-style "people with link can...") per server
- access-via-code recorded separately, deleted when share code goes away
- no expiry anywhere, but single link can be easily revoked

I think both make sense and are similarly simple. The first is more of an "invite code" model, where you are inviting folks to a server to stay, whereas the second is a "people with the link" model. The latter is easier to revoke permissions, while the former is easier to have short-term invite links. I don't know what folks would find most comfortable / familiar.

Another _possibly_ simpler approach from the Hub perspective would be to implement a generic API for _roles_ (#3858), then a 'sharing service' could layer on top the higher-level representation of what a 'share' is. That's not simpler for _me_, who needs to implement both, but it _might_ be simpler to implement in the Hub (roles in general present their own challenges if they can be user-created). This is ultimately more complex _overall_.

minrk commented:
Updating notes from today's meeting:

- reiterating keeping it simple
- that likely means only supporting invite codes permanently exchanged for 'shares'
- single permissions per `(user, server)`, so adding permissions is _updating_ this field, not adding a record
- no expiration for now, that can come later (invites expire, not permissions)

I think this is simple and clear enough that I can get something working before needing to bug folks for more feedback.

minrk commented:
Still some work to do, but I now have sharing fully working where a user can grant and revoke access to their server to another user.

Still todo:

- update doc 
- update REST API spec
- test coverage
- share code API

The fact that shares can be referenced both by the server and the user it's granted to has been surprisingly hard to make sense of in the REST API.

As it is now, I have:

By server:

- grant permission: POST /shares/:user/:server (shared-with, scopes in body)
- revoke permission: PATCH /shares/:user/:server (shared-with, scopes in body) (not delete because delete can't have body and don't have a URL that fully specifies owner, server, and shared_with)
- revoke _all_ shares for a server DELETE /shares/:user/:server

By user/group:

- list shared-with-me (user): GET /api/users/:name/shared
- list shared-with-me (group): GET /api/groups/:name/shared
- opt-out of share: DELETE /api/users/:name/shared/:owner/:server (separate endpoint from PATCH above because this requires different permissions - users can always un-share with themselves)

minrk commented:
Making progress, working on sharing codes, and came up with another question:

Is a sharing code more like a _token_, which should be stored hashed and irretrievable (must issue new code if you need access) or a more permanent, _retrievable_ value.

I think how folks use are expected to use them informs this. If we're thinking of slack or otherwise invitation-style links, I think the token format is better.

If we're thinking of Google docs style "share via code", then being able to retrieve the code so you don't reissue codes too often makes sense.

Given that share codes are immediately _exchanged_ for sharing permission, and permissions granted _via_ the code cannot be readily tracked or revoked, I'm inclined to go with the token approach.

minrk commented:
Still quite a few tests to write and things to update, but now I've tested the full process of issuing and accepting a new sharing code, and things are reasonably complete (not finished, but _complete_).

For this PR, what I plan to finish:

- documentation explaining how it works and how to use it
- sharing with individuals via API
- list/modify/revoke shares via API
- issue/list/revoke share codes via API
- accept share code invitations via human page (the only HTML endpoint I plan for this PR)

What I _don't_ plant to do in this PR:

- UI for managing shares
 

UI is its own whole can of worms, and I'd like to be able to make a 5.0 release with sharing made _possible_ via the API, and consider UI on a longer timescale (and via extensions in the meantime)

minrk commented:
Still need to work on docs and test coverage, but this is pretty thoroughly working, now.

minrk commented:
Once I have the API schema, reference docs, and an example in place, I'll ask folks for reviews.

minrk commented:
This is working, and I've updated the REST API and added an example where a user can issue sharing codes from a notebook granting access to another user.

I still need to expand a bit of test coverage and work out a couple of lazy-loading things, but this is just about ready.

minrk commented:
Test coverage is now pretty thorough, and tests are passing locally. I think this is ready for review!

Key things to look at:

1. reference documentation for how it does/should work
2. the example showing the process of granting access in the absence of a proper JupyterLab extension
3. the REST API schema

I think I want to make one more _tutorial_ narrative doc that I have in process, and then I'd say this is ready for merge (pending tests passing on machines other than mine, etc.).

minrk commented on jupyterhub/scopes.py:
```diff
@@ -646,68 +698,112 @@ def _resolve_requested_scopes(requested_scopes, have_scopes, user, client, db):
+
 
+@lru_cache_key(_has_scope_key)
+def has_scope(scope, have_scopes, *, post_filter=False, db=None):
```
Since this utility is important and used often in the new code, I've promoted checking for a scope to a simple public API

minrk commented on jupyterhub/scopes.py:
```diff
@@ -188,11 +217,12 @@ def groups_for_user(username):
+            .join(orm.User.groups)
+            .filter(orm.User.name == username)
+        )
+        return {row[0] for row in group_query}
```
Some query optimizations, shouldn't have an effect on results

minrk reviewed (commented):
Fixed sqlalchemy 1.4 compatibility and added [sharing tutorial](https://jupyterhub--4594.org.readthedocs.build/en/4594/tutorial/sharing.html). I don't have any further plans for this PR other than some more self-reviewing.

minrk commented:
Summarizing some feedback from yesterday's meeting:

> it seems plausible that a deployment might want to explicitly enable share codes or share-by-name without the other. If this is not in the first release (they use the same scope), is it a backward-compatible change to add it?

Answer: it can be fully backward compatible via `subscopes` (`read:users` is a subscope of `users` , so having `users` means you also have the `read:users` scope). So if we deem this desirable in a future release, we add two new scopes, e.g. `share-by-name` and `share-codes`, and make them subscopes of `shares`.

> Limiting codes to produce one-time-use. Or even broader limitations like restricted use?

We definitely can implement codes that can only be used once. This would be a small chage. Adding this feature after release would require an additional column in the ShareCodes table, so would require a minimally disruptive db upgrade. Current policy dictates that as a major release, despite being a tiny, fully backward-compatible change.

sgaist commented on jupyterhub/tests/test_shares.py:
```diff
@@ -0,0 +1,1448 @@
+    share_scopes,
+    status,
+):
+    def _expand_scopes(scope_str):
+        return [
+            s.replace("USER", user.name)
+            .replace("SERVER", user.name + "/")
+            .replace("SHARE_WITH", share_with.name)
+            for s in scope_str.split(",")
+        ]
```
Same implementation again

sgaist commented on requirements.txt:
```diff
@@ -10,6 +10,7 @@ packaging
 pamela>=1.1.0; sys_platform != 'win32'
 prometheus_client>=0.5.0
 psutil>=5.6.5; sys_platform == 'win32'
+pydantic>=2
```
Since it comes into a new major release, I would say it is fine to keep things simple and update pydantic.

consideRatio replied:
Let's go with >2 straight away as its a major release. I see that it has been around since Jun 30, 2023, so its not super new. If there is another software installed next to jupyterhub, it may very well have a release out that supports pydantic 2 at this point.

sgaist reviewed (commented):
There are several repetition in my comments in the tests code but as noted in one of them, it was to ensure we would catch them all.

minrk commented on docs/source/_static/rest-api.yml:
```diff
@@ -858,9 +862,305 @@ paths:
             - groups
       x-codegen-request-body-name: body
+
+  /shares/{owner}:
```
Yes, it's here and documented above as `/users/{name}/shared` for both users and groups.

minrk commented on docs/source/_static/rest-api.yml:
```diff
@@ -1380,6 +1769,95 @@ components:
+          type: array
+          items:
+            type: string
+        group:
+          description:
+            the group being shared with (exactly one of 'user' or 'group'
+            will be non-null, the other will be null)
+          type: object
+          nullable: true
+          properties:
+            name:
+              type: string
+        user:
+          description:
+            the user being shared with (exactly one of 'user' or 'group'
+            will be non-null, the other will be null)
+          type: object
+          nullable: true
+          properties:
+            name:
+              type: string
```
oneOf requires additional schemas for the mutually exclusive groups. I think this is clearer than adding the abstractions required to make the schema enforce mutual exclusivity, since it doesn't communicate it any better than the prose, and this is exclusively intended for docs generation.

minrk commented on docs/source/_static/rest-api.yml:
```diff
@@ -858,9 +862,305 @@ paths:
+            - read:users:name
+            - read:groups:name
+    patch:
+      summary: Revoke shared access
```
It's not modify, since it's explicitly and exclusively revocation. But on a per-user per-scope basis. But it does point to the fact that the delete method, which does exist and revokes all shared access to the server is missing here. I've added that and expanded the description of patch (summaries must stay very short to render nicely)

minrk commented on docs/source/_static/rest-api.yml:
```diff
@@ -858,9 +862,305 @@ paths:
+        - oauth2:
+            - read:shares
+
+  /shares/{owner}/{server_name}:
```
Thanks, added.

minrk commented on docs/source/reference/sharing.md:
```diff
@@ -0,0 +1,364 @@
+
+    },
+    "scopes": ["access:servers!server=username/servername"],
+    "user": { # or null
```
could be. The request is JSON, so null would be correct, but this is highlighted Python output.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+        return scopes
+
+
+class ShareGrantRequest(BaseShareRequest):
+    """Validator for requests to grant sharing permission"""
+
+    # directly granted shares don't expire
+    # since Shares are _modifications_ of permissions,
+    # expiration can get weird
+    # if it's going to expire, it must expire in
+    # at least one minute and at most 10 years (avoids nonsense values)
+    # expires_in: conint(ge=60, le=10 * 525600 * 60) | None = None
+    user: Optional[str] = None
+    group: Optional[str] = None
+
+    @model_validator(mode='after')
+    def user_group_exclusive(self):
+        if self.user and self.group:
+            raise ValueError("Expected exactly one of `user` or `group`, not both.")
+        if self.user is None and self.group is None:
+            raise ValueError("Specify exactly one of `user` or `group`")
+        return self
+
+
+class ShareRevokeRequest(BaseShareRequest):
+    """Validator for `revoke` field of requests to revoke shares"""
+
+    user: Optional[str] = None
+    group: Optional[str] = None
+
+    @model_validator(mode='after')
+    def user_group_exclusive(self):
+        if self.user and self.group:
+            raise ValueError("Expected exactly one of `user` or `group`, not both.")
+        if self.user is None and self.group is None:
+            raise ValueError("Specify exactly one of `user` or `group`")
+        return self
+
```
Yes. They didn't before because `expires_in` used to be present, but now they are indeed the same.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+
+
+class ShareRevokeRequest(BaseShareRequest):
+    """Validator for `revoke` field of requests to revoke shares"""
```
Good catch, used to be a single `modify` endpoint with grant/request fields, but now separate endpoints. Fixed.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+            return user.orm_spawners[server_name]
+        if raise_404:
+            raise web.HTTPError(404, f"No such server: {user_name}/{server_name}")
+        else:
+            return None
```
Doesn't really matter, I just think it's clearer sometimes to have two mutually exclusive branches, rather than _implicit_ mutual exclusivity by having one terminating branch. Either is fine.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+
+    @needs_scope("read:users:shares")
+    def get(self, user_name):
+        query = self._init_share_query()
```
makes sense.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+        if user is None:
+            raise web.HTTPError(
+                404,
+                f"No such share for user {user_name} on {owner_name}/{server_name}",
```
Yes, but the error shouldn't reveal that information. The error a user sees for a user not existing and the share not existing should be the same.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+
+        raises 404 if not found
+        """
+        group = self.find_group(group_name)
```
This is that check. `find_group` raises 404 if not found.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+    async def post(self, user_name, server_name=None):
+        """POST grants permissions for a given server"""
+
+        if server_name is None:
+            # only GET supported `/shares/{user}` without specified server
+            raise web.HTTPError(405)
```
it must be optional to support a single RequestHandler class on two url patterns.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+
+    @needs_scope('shares')
+    async def patch(self, user_name, server_name=None):
+        """PATCH revokes single shares for a given server"""
```
clarified the docstring that it revokes _permissions_ from single shares, not necessarily shares themselves

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+
+    @needs_scope('shares')
+    async def delete(self, user_name, server_name=None):
+        if server_name is None:
+            # only GET supported `/shares/{user}` without specified server
+            raise web.HTTPError(405)
```
Addressed above.

minrk commented on jupyterhub/apihandlers/shares.py:
```diff
@@ -0,0 +1,598 @@
+        share_id = self.get_argument("id", None)
+        spawner = self._lookup_spawner(user_name, server_name)
+        if code:
+            # delete one code by code
```
clarified

minrk commented on jupyterhub/scopes.py:
```diff
@@ -883,21 +981,53 @@ def _auth_func(self, *args, **kwargs):
                     resource_value = bound_sig.arguments[resource_name]
                     s_kwargs[resource] = resource_value
+
+            if "server" in s_kwargs:
+                # merge user_name, server_name into server=user/server
+                if "user" not in s_kwargs:
+                    raise ValueError(
+                        "Cannot filter on 'server_name' without 'user_name'"
+                    )
+                s_kwargs["server"] = f"{s_kwargs['user']}/{s_kwargs['server']}"
+                s_kwargs.pop("user")
```
A reasonable suggestion. I'd rather not add that level of refactor to this PR, but if you want to open a separate Issue suggesting that, I think it makes sense.

sgaist replied:
Sure thing !

The ticket is here #4681

minrk commented on jupyterhub/tests/test_orm.py:
```diff
@@ -495,6 +497,73 @@ def test_group_delete_cascade(db):
+def test_share_user(db):
+    user1 = orm.User(name='user1')
+    user2 = orm.User(name='user2')
+    group1 = orm.Group(name='group1')
+    group2 = orm.Group(name='group2')
```
copy pasta starting from another similar test. Removed.

minrk commented on jupyterhub/tests/test_shares.py:
```diff
@@ -0,0 +1,1448 @@
+        expected_shares.append(f"sc_{code.id}")
+
+    expected_shares = sorted(expected_shares)
+    if level == 'user':
+        path = f"/share-codes/{user.name}"
+    else:
+        path = f"/share-codes/{user.name}/"
```
the server is part of the path. The default server name is the empty string, so the url `/share-codes/user` gets shares for a user, while `/share-codes/user/` gets shares for the server named `''`.

sgaist commented on docs/source/_static/rest-api.yml:
```diff
@@ -737,29 +694,85 @@ paths:
+        204:
+          description: |
+            Permission has been revoked,
+            the user no longer has access to the server.
```
```suggestion
            the group members no longer have access to the server.
```

sgaist reviewed (approved):
Some additional minor stuff otherwise it's looking pretty good.

This time with suggestions to make things easier.

consideRatio commented:
(Merge conflict resolution needed)

minrk commented:
conflict resolved

sgaist reviewed (approved):
Just a oneOf suggestion but I would not consider it blocking.

The patch looks good.

consideRatio reviewed (commented):
Amazing work on this @minrk!!! I've started reviewing, but I'll need to break up my efforts in parts. I've started with reading through the reference docs on sharing. These are my notes.

# Sharing reference docs

## [Sharing access to user servers (intro)](https://jupyterhub--4594.org.readthedocs.build/en/4594/reference/sharing.html)

- Intro to help understanding the meaning of access
  
  The title includes "access", but what does that mean? My current understanding is that having "access" historically has meant that when a user has has visited a user server (everyone can, proxy isn't authenticated), and the user server asks JupyterHub who the user is, then JupyterHub will provide user information only if the user has "access".

  How does the historical access granting relate to the fine grained access that could be supported by the user server software? This kind of question may be better explained an explanation docs section or similar, but its context that I think is helpful to have when reading this reference text.

  - [x] Consider if a clarification of access can be described someplace, maybe in a separate explanation docs or put in this reference page but isolated under a header to more easily be refactored later if its considered explanation.

## [The share model](https://jupyterhub--4594.org.readthedocs.build/en/4594/reference/sharing.html#the-share-model)

- [x] Link out

  The example is hardcoded atm, which is acceptable, but I think it should then include some connection to verify that the hardcoded part isn't outdated for example.
  
  Maybe this could also link out to a REST API reference docs response, returning these kinds of model instances?

  I saw in [the REST API reference](https://jupyterhub--4594.org.readthedocs.build/en/4594/reference/rest-api.html#/default/get_shares__owner___server_name_) for example, which notably also includes "active" in server btw, omitted from the hardcoded example.

## [Share via invitation code](https://jupyterhub--4594.org.readthedocs.build/en/4594/reference/sharing.html#share-via-invitation-code)

- Single user or multiple use share codes?
  Is the share code usable once or multiple times until it expires? The text writes about "one or mor users" but also that a code is exchanged (as compared to "passed" or similar), and that hints to me that its only valid for its first use.
  - [ ] Clarify this in docs

- Share codes for a group?
  Can a share code only be accepted by a user for a user? Is there a share code interaction making a group be granted permissions?
  - [ ] Clarify this in docs

- The docs include a TODO ("TODO: which may or may not be running!")
  - [ ] Cleanup todo note

- Example on listing share codes doesn't include "expires_at" or similar, but should right?
  - [ ] Ensure share code examples captures an "expires" field that reasonably should be there

## The share code model

- [ ] There isn't a share code model heading (like for the share model heading)

  In [the GET share code's API response](https://jupyterhub--4594.org.readthedocs.build/en/4594/reference/rest-api.html#/default/get_share_codes__owner___server_name_) I see details about the share code's model that includes "created_at" and "expires_at", things I expected to be included but didn't see mentioned.

minrk commented:
Great questions, thanks @consideRatio! I'll work on answers.

minrk commented on docs/source/_static/rest-api.yml:
```diff
@@ -859,9 +863,324 @@ paths:
+      parameters:
+        - $ref: "#/components/parameters/sharedServerOwner"
+        - $ref: "#/components/parameters/sharedServerName"
+        - in: query
+          name: code
+          description: the share code to revoke
+          schema:
+            type: string
+        - in: query
+          name: id
+          description: the id of the share code to revoke
+          schema:
+            type: string
```
Since this file is solely for the purposes of generating docs for humans, I don't think there's anything to change here. The renderer we currently use (swaggerUI) doesn't handle this multi-level schema appropriately, nor does it communicate the mutual exclusivity when configured this way.

This PR currently (clearly documents code and id as separate parameters, and that exactly one should be used):

<img width="781" alt="Screenshot 2024-01-29 at 11 10 40" src="https://github.com/jupyterhub/jupyterhub/assets/151929/cb7fc7f6-6202-4a7e-8bfd-0d242dab80f7">

With the suggested change (less clear, discards exclusivity information):

<img width="762" alt="Screenshot 2024-01-29 at 11 09 57" src="https://github.com/jupyterhub/jupyterhub/assets/151929/93c447a8-a68d-4687-b889-41afe7ae807a">

sgaist replied:
Agreed, it's more confusing than anything else.

Thanks for the images.

consideRatio commented:
Review continuation, more to follow as well.

# [Tutorial docs](https://jupyterhub--4594.org.readthedocs.build/en/4594/tutorial/sharing.html) review notes

- [ ] Retitle ["Step 0"](https://jupyterhub--4594.org.readthedocs.build/en/4594/tutorial/sharing.html#step-0-enable-sharing-admin)?
  To lead with "Step 0:" to some degree indicated other steps, what do you think about retitling to pre-requisite or similar?
- [ ] Clarify "sharing enabled" and what reverting it leads to
  Will reverting the "shared enabled" step mean that users who have been granted shares no longer can use that access, or only that no further sharing can be done? It would be good to have it explicit what is enabled - the creation of shares, or the creation and use of shares.

## [Get a token](https://jupyterhub--4594.org.readthedocs.build/en/4594/tutorial/sharing.html#get-a-token)

- [ ] Consider input of token for code snippet
  What do you think about augmenting the code snippet example to include a `import getpass` and `token = getpass.getpass('JupyterHub API token: ')`?

## [Issue a sharing code](https://jupyterhub--4594.org.readthedocs.build/en/4594/tutorial/sharing.html#issue-a-sharing-code)

- [ ] Indicate that a code can be exchanged multiple times

consideRatio commented:
## Other non-code review notes

- On permissions granularity and defaults
  
  Seems ok to me!

- Database schema, what info is stored etc

  No changes suggested, but reflected like below.

  - **share model**
    I've deliberated on shares being tied to (user, server), an an alternative like (user, server, creation-uid) where creation-uid is generated on creation by perhaps being a hash of all relevant fields to ensure no duplicate entries are created. No change suggested, but I was thinking with such system, it may be possible to transition to time-outable shares etc. But, how important is this? I figure one may want to allow temporary read-only permissions or similar to view a session without letting someone view future sessions.
  - **share-code model**
    I've also considered if this should already or in the future include a maximum exchange number or similar, but I think the nice thing is that this behavior and check can be added retroactively without needing to be part of this PR. So, I think all is good for now.

- **Share codes for groups?**
  I've understood that there is a way for users to accept shares via share codes, but is this workflow unique for users, or is it something that could/should apply for groups as well? For example, could a user accept a share as a group its part of etc?
  
  I think the action point is to detail if the accept form and/or share codes are for users only atm if that is a known kind of constraint.

---

Okay, time for code review now, then I'm done with the initial review pass!

consideRatio commented on jupyterhub/handlers/pages.py:
```diff
@@ -552,6 +552,62 @@ def sort_key(client):
+            username=self.current_user.name,
+        )
+
+        html = await self.render_template(
+            'accept-share.html',
+            code=code,
+            owner=share_code.owner,
+            spawner=share_code.spawner,
+            scope_descriptions=scope_descriptions,
+            next_url=next_url,
+        )
```
Should this form be informed about if the server is started/ready etc?

minrk replied:
that's a good idea. I'll see how easy that is to add.

minrk replied:
Now looks like this if the server isn't running:

<img width="808" alt="Screenshot 2024-02-06 at 12 29 22" src="https://github.com/jupyterhub/jupyterhub/assets/151929/55d9ea10-8fec-4ad6-a138-c86304f2a08c">

consideRatio commented on jupyterhub/scopes.py:
```diff
@@ -143,6 +143,36 @@
     'access:services': {
         'description': 'Access services via API or browser.',
     },
+    'users:shares': {
+        'description': "Read and revoke a user's access to shared servers.",
+        'subscopes': [
+            'read:users:shares',
+        ],
+    },
```
Isn't being granted `users:shares` being granted the ability to read/create/revoke any user's shares, and not just a user and not just read/revoke?

I was thinking maybe a filter is what limits this permission to a speific user, but this permission is described without a filter applied as a whole now.

minrk replied:
Hm, I'm never quite satisfied with scope descriptions and combining with filters. All scopes are described assuming no filter, and then filters limit e.g. _which_ user can be acted upon.

Having this scope means you can take the _action_ "read and revoke a user's access to shared servers", which is the prose representation of `GET|DELETE /api/users/:name/shared[/:server]`, while the filter specifies _which_ users you can take that action upon. If you have a clearer summary that remains concise, I think there's always room for improvement here.

consideRatio reviewed (approved):
I don't see a blocker. I've not reviewed all changes in detail due to limited capacity =/

I've glossed over changes in orm.py and scopes.py, and hardly looked at shares.py. Nothing seems off from what I've seen in all other code sections.

Amazing effort into this @minrk! There are some unaddressed reivew points above, but none is critical so I'll leave it to you to decide if you want to take action on them before going for a merge.

minrk commented:
> I've deliberated on shares being tied to (user, server), an an alternative like (user, server, creation-uid) where creation-uid is generated on creation by perhaps being a hash of all relevant fields to ensure no duplicate entries are created. No change suggested, but I was thinking with such system, it may be possible to transition to time-outable shares etc. But, how important is this? I figure one may want to allow temporary read-only permissions or similar to view a session without letting someone view future sessions.

Implementation-wise, it would just be like other tables (User, group, etc.), where there's a unique `id` primary key, and drop the restriction on uniqueness of `(user, server)`. This was in fact the original design, where you could grant expiring permissions. Short-term read-only access was indeed the use case I had in mind. Where things got really hairy was when we started to talk about _revoking_ permissions, especially when a user can be granted the same permission multiple times with multiple expiration dates. What does a revocation look like when there are multiple Share objects per `(user, server)`? It is revoking a whole 'share', which may or may not revoke permissions? Updating a single record in-place makes this all a lot simpler, but at the expense of short-term expiring grants not being feasible. This will be a hard thing to change if we want to create expiring shares, because several assumptions in the API change - e.g. `GET /user/:name/shared` would have to change from a single Share model to a list of shares.

> I've understood that there is a way for users to accept shares via share codes, but is this workflow unique for users, or is it something that could/should apply for groups as well? For example, could a user accept a share as a group its part of etc?

share codes are just for users. Sharing with groups can only be done explicitly by name. I'm not sure it makes sense use-case-wise for a user to be able to accept a code on behalf of a group. That's essentially delegating sharing permission, where the owner of the server is granting anyone with the code permission to share the owner's server with a group by name. It's not impossible to implement, and wouldn't require any changes to the database, but I'd like to see a real use case requirement first.

> Permission issues with custom read-only scopes?

It is conceivable to have two images where one supports read-only scopes and the other doesn't, and the result would likely be ignoring scoped permissions and granting full access. I think this ought to be the responsibility of admins. E.g. if admins (and admins are _required_ for read-only access, since they must define the custom scopes), then they shouldn't allow launching images that don't have read-only support. Allowing user control over the server environment (e.g. user-controlled images) means granting users the power to completely eliminate all security protections.

For server-session-scoped permission grants, this is feasible _globally_, but not on a per-permission basis because of the single `(user, server)` record. That choice - to make it easy to list, manage, and revoke permissions - means that all permission revocation must be manual (just like it is for ~all permission-granting I've ever seen, such as github, google docs, etc.). Also note that _deleting_ a server, which is distinct from _stopping_ it, does indeed revoke all share permissions because the relationship is removed. It is only typical to delete _named_ servers:

<img width="695" alt="Screenshot 2024-02-06 at 11 10 59" src="https://github.com/jupyterhub/jupyterhub/assets/151929/004223db-ed7b-4946-a660-9307252d8fe3">

I'll address other comments in-line.

Thanks for reviewing!

minrk commented:
>  but "active" isn't?

this was my confusion mixing internal variables with the rest API. `.active` is a Python variable, but not in the REST API at all (`active = ready or pending`, by definition). But I see that it's _documented_ in the REST API, despite not being present! I've updated the rest docs to reflect reality.

minrk commented:
> In [this example response](https://jupyterhub--4594.org.readthedocs.build/en/4594/reference/rest-api.html#/default/get_share_codes__owner___server_name_) I see an expires_at an created_at equal to each other, does that imply that it won't expire? It seems a bit more natural that the field expires_at wouldn't be included or that its extremely far into the future or something.

This is because unless we specify individual examples for everything, all timestamp fields use the same value (the current time). We can specify manual examples. I've expanded the description to state that expires_at is always in the future, and `null` means no expiration.

consideRatio commented:
**1. Warn that a read-only constraint is enforced in user environments**

> > Permission issues with custom read-only scopes?
> 
> It is conceivable to have two images where one supports read-only scopes and the other doesn't, and the result would likely be ignoring scoped permissions and granting full access. I think this ought to be the responsibility of admins. E.g. if admins (and admins are _required_ for read-only access, since they must define the custom scopes), then they shouldn't allow launching images that don't have read-only support. Allowing user control over the server environment (e.g. user-controlled images) means granting users the power to completely eliminate all security protections.

I think this info is critical info that should be surfaced very intentionally to admins.

- [ ] Check if this info can be surfaced even more clearly to avoid security issues

minrk commented:
> I think this info is critical info that should be surfaced very intentionally to admins.

I'll think about where to highlight it. I don't think it's particularly relevant to sharing, so I wouldn't add anything to the sharing docs, but it is relevant to the custom_scopes definitions. The user-control over the environment is discussed already in [the security doc](https://jupyterhub.readthedocs.io/en/stable/explanation/websecurity.html), with the statement that:


> JupyterHub administrators must ensure that:
> A user does not have permission to modify their single-user notebook server,

including installing packages, modifying $PATH, etc., all of which enable users to compromise security of their own servers.

consideRatio reviewed (approved):
Amazing effort into this @minrk!!! 👍 For squash/merge/get this pr into main branch!

sgaist reviewed (approved):
Amazing work !

minrk commented:
Thanks so much for the reviews!

krassowski commented:
Very happy to see this merged :tada: What is the release plan for this feature? Is it backwards compatible and could land in say 4.1 or will it need to wait for 5.0?

minrk commented:
This is the main feature for 5.0, which I plan to start prereleases for shortly. Working on one more fix before that can start.
